src/Schema/Directives/CanDirective.php

```diff
@@ -82,22 +82,22 @@ public static function definition(): string
     public function handleField(FieldValue $fieldValue, Closure $next): FieldValue
     {
         $previousResolver = $fieldValue->getResolver();
+        $ability = $this->directiveArgValue('ability');
 
-        return $next(
-            $fieldValue->setResolver(
-                function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo) use ($previousResolver) {
-                    $gate = $this->gate->forUser($context->user());
-                    $ability = $this->directiveArgValue('ability');
-                    $checkArguments = $this->buildCheckArguments($args);
+        $fieldValue->setResolver(
+            function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo) use ($ability, $previousResolver) {
+                $gate = $this->gate->forUser($context->user());
+                $checkArguments = $this->buildCheckArguments($args);
```

Comment:
`buildCheckArguments` is querying directive args as well. We have to pass `injectArgs`/`args` to it.

Or it may be a better idea to cache arguments in `directiveArgValue`/`directiveHasArgument` functions - so after the first query to AST, they will be saved locally. Or even load all directive args on the first query. The directive code will be simpler and easier to understand, while performance should be almost the same.

Reply:
The difficulty with this particular code is that PHP has no easy way of representing `undefined`, and there is a difference between:

```graphql
@can(args: null) // Should call check($ability, [null])
@can // Should call check($ability, [])
```

The `directiveArgLookup` is not that expensive, so I think we can just leave this as is.

Reply:
Centralized loading of directive args might be a good idea, I thought about implementing a generalized solution that runs validation during the schema compilation.

Reply:
Here is what I meant to do:
https://github.com/nuwave/lighthouse/pull/1707
Args are loaded once and can be swiftly used in any place later.

Reply:
Since #1707 landed, this no longer costs much performance, so I am going to leave it.